crates/stackable-certs/src/ca/ca_builder.rs

```diff
@@ -0,0 +1,240 @@
+use bon::Builder;
+use rsa::pkcs8::EncodePublicKey;
+use snafu::{ResultExt, Snafu};
+use stackable_operator::time::Duration;
+use tracing::debug;
+use x509_cert::{
+    builder::{Builder, CertificateBuilder, Profile},
+    der::{DecodePem, referenced::OwnedToRef},
+    ext::pkix::AuthorityKeyIdentifier,
+    name::Name,
+    serial_number::SerialNumber,
+    spki::SubjectPublicKeyInfoOwned,
+    time::Validity,
+};
+
+use super::CertificateAuthority;
+use crate::{
+    CertificatePair,
+    ca::{DEFAULT_CA_VALIDITY, PEM_LINE_ENDING, SDP_ROOT_CA_SUBJECT},
+    keys::CertificateKeypair,
+};
+
+/// Defines all error variants which can occur when creating a CA
+#[derive(Debug, Snafu)]
+pub enum CreateCertificateAuthorityError<E>
+where
+    E: std::error::Error + 'static,
+{
+    #[snafu(display("failed to parse validity"))]
+    ParseValidity { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to parse \"{subject}\" as subject"))]
+    ParseSubject {
+        source: x509_cert::der::Error,
+        subject: String,
+    },
+
+    #[snafu(display("failed to create signing key pair"))]
+    CreateSigningKeyPair { source: E },
+
+    #[snafu(display("failed to serialize public key as PEM"))]
+    SerializePublicKey { source: x509_cert::spki::Error },
+
+    #[snafu(display("failed to decode SPKI from PEM"))]
+    DecodeSpkiFromPem { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to parse AuthorityKeyIdentifier"))]
+    ParseAuthorityKeyIdentifier { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to create certificate builder"))]
+    CreateCertificateBuilder { source: x509_cert::builder::Error },
+
+    #[snafu(display("failed to add certificate extension"))]
+    AddCertificateExtension { source: x509_cert::builder::Error },
+
+    #[snafu(display("failed to build certificate"))]
+    BuildCertificate { source: x509_cert::builder::Error },
+}
+
+/// This builder builds certificate authorities of type [`CertificateAuthority`].
+///
+/// Example code to construct a CA:
+///
+/// ```no_run
+/// use stackable_certs::{
+///     keys::ecdsa,
+///     ca::{CertificateAuthority, CertificateAuthorityBuilder},
+/// };
+///
+/// let ca: CertificateAuthority<ecdsa::SigningKey> = CertificateAuthorityBuilder::builder()
+///     .build()
+///     .build_ca()
+///     .expect("failed to build CA");
+/// ```
+#[derive(Builder)]
+pub struct CertificateAuthorityBuilder<'a, SKP>
```

Comment:
I thin this should be:

```suggestion
pub struct CertificateAuthority<'a, SKP>
```

So when we call `build()`, we get back a CA and not a CA Builder.

_Note, the Builder macro would otherwise generate `CertificateAuthorityBuilderBuilder` (ie: it generates the Builder, not us)._

Reply:
To illustrate further:

- `CertificateAuthority::builder() -> CertificateAuthorityBuilder`
- `CertificateAuthorityBuilder::build() ->` a valid `CertificateAuthority`

Reply:
Oh, but we already have a struct: `CertificateAuthority<SK>` at https://github.com/stackabletech/operator-rs/blob/88d850d14abe56df13af230c2b74eaa5cb5cfd1e/crates/stackable-certs/src/ca/mod.rs#L19

in my mind, that one should be the one that derives Builder. Otherwise we need to revise what each struct is, because the current `CertificateAuthorityBuilder` is not a builder (per previous message above).